packages/pkl.lua/lua.pkl

```diff
@@ -294,3 +332,560 @@ class Renderer extends ValueRenderer {
+    else (this) {
+      map = super.map.put(super.key!!, value)
+      key = null
+      op = null
+      negate = null
+    }
+
+  // used for _ENV[0]=value expressions when useDynamic is true
+  function add(value, _): ParseState =
+    (this) {
+      list = super.list.add(value)
+      key = null
+      op = null
+      negate = null
+    }
+
+  // only use this for [key]= keys
+  function setKey(value, _): ParseState =
+    (this) {
+      key = value
+      op = "key"
+      negate = null
+    }
+
+  function negate(): ParseState =
+    (this) { negate = !(super.negate ?? false) }
+
+  function isListIndex(key): Boolean =
+    // note: Lua indexes are 1-based
+    key is Int && key == list.length + 1
+}
+
+local class ChildParseState extends ParseState {
+  parent: ParseState
+  brace: RegexMatch // token for opening {, used for error reporting
+  op: ("["|"key"|"]"|"="|"value")? // op is "["|"value"? iff key is null
+  mapStart: RegexMatch? // non-null if !map.isEmpty
+  listStart: RegexMatch? // non-null if !list.isEmpty
+
+  function toValue(useDynamic: Boolean, source: String, uri: Uri?): Dynamic|Mapping|Listing =
+    if (useDynamic) (map.toDynamic()) {
+      ...list
+    } else if (map.isEmpty) list.toListing()
+    else if (list.isEmpty) map.toMapping()
+    else throwError2("table has both list elements and map entries", source, uri, mapStart!!, "first map entry", listStart!!, "first list entry")
+
+  function put(value, token: RegexMatch, _): ChildParseState =
+    if (isListIndex(key)) add(value, token)
+    else (this) {
+      map = super.map.put(super.key!!, value)
+      key = null
+      op = "value"
+      negate = null
+    }
+
+  function add(value, token: RegexMatch): ChildParseState =
+    (this) {
+      list = super.list.add(value)
+      key = null // key could be [int]
+      op = "value"
+      negate = null
+      listStart = super.listStart ?? token
+    }
+
+  function setKey(value, token: RegexMatch): ChildParseState =
+    (this) {
+      key = value
+      op = "key"
+      negate = null
+      mapStart =
+        // don't update mapStart if this will become a listing element
+        if (isListIndex(value)) super.mapStart
+        else super.mapStart ?? token
+    }
+}
+
+// endregion
+// region Tokens
+
+// Regex that matches a single Lua token, or an invalid character.
+// This regex assumes line endings have already been normalized, so no carriage returns exist.
+// Error states:
+// - Group 2 is ""
+// - Group 3 is not "\""?
+// - Group 4 is not "'"?
+// - Group 6 is ""
+// - Last group is non-null
+local const tokenRegex: Regex = Regex(##"""
+  (?x-uU)
+    --(?:               # comment
+      \[(?<eq1>=*)\[    # long comment (equals are GROUP 1)
+      (?>(?>
+        [^]]
+      | ](?!\k<eq1>])
+      )*)
+      (]\k<eq1>]|\z)    # match end or EOF (GROUP 2)
+    | .*                # short comment
+    )
+  | [\w&&\D]\w*         # identifier
```

Comment:
`\D` is any character that is not a digit. This is probably not what you meant?  This matches `*foo` for example.

Reply:
`[\w&&\D]` means the intersection of `\w` and `\D`, so the pattern `[\w&&\D]\w*` means "a word character that's not a digit, followed by zero or more word characters". Since I've disabled unicode character classes, this is equivalent to writing `[a-zA-Z_][a-zA-Z0-9_]*`.

I originally had this as the definition for `identifierRegex` too, but I changed it there to get rid of a false-positive warning claiming that `[\w&&\D]` can be simplified to `\w&&\D`. Perhaps I should have changed it here too for consistency.

Reply:
Ah, I see; read this as a character class at first. My bad!